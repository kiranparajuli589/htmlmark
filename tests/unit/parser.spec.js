const lexer = require("../../lib/lexer")
const parser = require("../../lib/parser")
const { commonTokensList } = require("../fixtures/commTokens")


function toHtml(text) {
  return parser(lexer(text))
}

describe("lexer", () => {
  describe("newline", () => {
    it.each([
      [""],
      ["\n"],
      ["", "\n"],

    ])("should not include the top empty lines", (lines) => {
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
    it.each(["", "\n"])("should include the empty line after some content", (eLine) => {
      const lines = [
        "some plain text",
        eLine,
      ]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
    it("should combine multiple consequest new lines to a single one", () => {
      const lines = [
        "some plain text",
        "\n",
        "",
        "some more plain text",
      ]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
  })
  describe("codeblock", () => {
    it("should parse the codeblock", () => {
      const lines = [
        "```js",
        "const a = 1",
        "```",
      ]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
    it("should parse the codeblock without the language", () => {
      const lines = [
        "```",
        "const a = 1",
        "```",
      ]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
    it("should parse the codeblock with multiple lines", () => {
      const lines = [
        "```js",
        "const a = 1",
        "const b = 2",
        "```",
      ]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
    it("should parse the codeblock with multiple lines and a newline", () => {
      const lines = [
        "```js",
        "const a = 1",
        "const b = 2",
        "",
        "const c = 3",
        "```",
      ]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
    it("should parse multiple consequetive codeblocks", () => {
      const lines = [
        "```js",
        "const a = 1",
        "```",
        "```js",
        "const b = 2",
        "```",
      ]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
    it("should cope with multiple items", () => {
      const lines = [
        "```js",
        "const a = 1",
        "```",
        "some people are funny",
      ]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
  })
  describe("common tokens", () => {
    it.each(commonTokensList)("should parse the common tokens", (line) => {
      const lines = [line]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
    it.each(commonTokensList)("should parse the common tokens with some space before", (line) => {
      const indent = "  "
      const lines = [indent + line]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
  })
  describe.skip("quote", () => {
    it.each([
      "> quote 1",
      "> > quote 2",
      "> > > quote 3",
      "> > > > quote 4",
      "> > > > > quote 5",
    ])("should detect the quote depth", (quote) => {
      const html = toHtml([quote])
      expect(html).toMatchSnapshot()
    })
    it.each([
      "> quote **one** with *two*",
      "  > quote `four` with ~~five~~",
      "> quote [link-title](link-url) with *two*",
    ])("should deep tokenize quote with zero depth", (line) => {
      const html = toHtml([line])
      expect(html).toMatchSnapshot()
    })
    it.each([
      "> > quote **one** with *two*",
      "  > > > quote `four` with ~~five~~",
      "> > > > quote [link-title](link-url) with *two*",
    ])("should deep tokenize quote with multiple depth", (line) => {
      const html = toHtml([line])
      expect(html).toMatchSnapshot()
    })
  })
  describe("list", () => {
    it("should tokenize a valid ordered list", () => {
      const lines = [
        "1. item **1**",
        "1. item [link](link-url)",
        "1. item 3 `code item`",
      ]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
    it("should tokenize a valid un-ordered list", () => {
      const lines = [
        "- item **1**",
        "- item [link](link-url)",
        "- item 3 `code item`",
      ]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
    it("should detech list indent", () => {
      const lines = [
        "  - item **1**",
        "  - item [link](link-url)",
        "  - item 3 `code item`",
      ]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
    it("indent should break the list", () => {
      const lines = [
        "  - item **1**",
        "  - item [link](link-url)",
        "- item 3 `code item`",
        "- item 4",
      ]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
    it("should tokenize list combination", () => {
      const lines = [
        "- one",
        "- two",
        "1. one",
        "2. two",
        "- [ ] c empty",
        "- [x] c checked",
        "1. [ ] c empty",
        "1. [x] c checked"
      ]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
  })
  describe("hr line", () => {
    it("should parse the hr line", () => {
      const lines = [
        "---",
      ]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
    it("should not allow multiple consecutive hr lines", () => {
      const lines = [
        "---",
        "---",
      ]
      const html = toHtml(lines)
      expect(html).toMatchSnapshot()
    })
  })
  describe("table", () => {
    describe("is not a table", () => {
      it("only header", () => {
        const html = toHtml([
          "| column 1 | column 2 |",
        ])
        expect(html).toMatchSnapshot()
      })
      it("only header and separator", () => {
        const html = toHtml([
          "| column 1 | column 2 |",
          "|---|---|",
        ])
        expect(html).toMatchSnapshot()
      })
      it("false separator", () => {
        const html = toHtml([
          "| column 1 | column 2 |",
          "| --- | --- |",
          "| row 1 c1 | row 1 c2 |",
        ])
        expect(html).toMatchSnapshot()
      })
      it("not consistent cell count", () => {
        const lines = [
          "| column 1 | column 2 |",
          "| --- |",
          "| row 1 c1 | row 1 c2 |",
        ]
        const html = toHtml(lines)
        expect(html).toMatchSnapshot()
      })
      it("other tokens in between 1", () => {
        const html = toHtml([
          "| column 1 | column 2 |",
          "",
          "|---|---|",
          "| row 1 c1 | row 1 c2 |",
        ])
        expect(html).toMatchSnapshot()
      })
      it("other tokens in between 2", () => {
        const html = toHtml([
          "| column 1 | column 2 |",
          "|---|---|",
          "",
          "| row 1 c1 | row 1 c2 |",
        ])
        expect(html).toMatchSnapshot()
      })
      it("other tokens in between 3", () => {
        const html = toHtml([
          "| column 1 | column 2 |",
          "|---|---|",
          "| row 1 c1 | row 1 c2 |",
          "",
          "| row 2 c1 | row 2 c2 |",
        ])
        expect(html).toMatchSnapshot()
      })
    })
    describe("table indent", () => {
      it("different indent header -> separator", () => {
        const html = toHtml([
          "| column 1 | column 2 |",
          "  |---|---|",
          "| row 1 c1 | row 1 c2 |",
        ])
        expect(html).toMatchSnapshot()
      })

      it("different indent header, separator -> body row", () => {
        const html = toHtml([
          "  | column 1 | column 2 |",
          "  |---|---|",
          "| row 1 c1 | row 1 c2 |",
        ])
        expect(html).toMatchSnapshot()
      })

      it("the indented table", () => {
        const html = toHtml([
          "  | column 1 | column 2 |",
          "  |---|---|",
          "  | row 1 c1 | row 1 c2 |",
        ])
        expect(html).toMatchSnapshot()
      })

      it("indent should break the table 1", () => {
        const html = toHtml([
          "  | column 1 | column 2 |",
          "  |---|---|",
          "  | row 1 c1 | row 1 c2 |",
          "| row 2 c1 | row 2 c2 |",
        ])
        expect(html).toMatchSnapshot()
      })
      it("indent should break the table 2", () => {
        const html = toHtml([
          "  | column 1 | column 2 |",
          "  |---|---|",
          "  | row 1 c1 | row 1 c2 |",
          "| column 1 | column 2 |",
          "|---|---|",
          "| row 1 c1 | row 1 c2 |",
        ])
        expect(html).toMatchSnapshot()
      })
    })

    describe("table cell count", () => {
      it("cell count should break the table 1", () => {
        const html = toHtml([
          "| column 1 | column 2 |",
          "|---|---|",
          "| row 1 c1 | row 1 c2 |",
          "| row 2 c1 |",
        ])
        expect(html).toMatchSnapshot()
      })
      it("cell count should break the table 2", () => {
        const html = toHtml([
          "| column 1 | column 2 |",
          "|---|---|",
          "| row 1 c1 | row 1 c2 |",
          "| column 1 |",
          "|---|",
          "| row 1 c1 |",
        ])
        expect(html).toMatchSnapshot()
      })
    })
    describe("table heading-body separator", () => {
      it.each([
        "|-------|-------|",
        "|:-------:|:---------:|",
      ])("should parse a table", (line) => {
        const html = toHtml([
          "| column 1 | column 2 |",
          line,
          "| row 1 c1 | row 1 c2 |",
          "| row 2 c1 | row 2 c2 |",
        ])
        expect(html).toMatchSnapshot()
      })
    })
  })
  describe("paragraph", () => {
    it.each([
      "some normal and **bold with * gem** but pure *italics* is alos there baby now ~~coming~~hola amigons~~strike~~ wooo lala what about blazing *********here baby ~~~~~~~~~baby `baby```[[[[[[[[[[[[[[[l]]]]int](href)[link](href)``````````",
      "a paragraph of <u>words</u> `first code` normal text here `code body` *first italics* here me crying *italic body* here me crying **first bolds** some normal again **bold body** [Kiran Parajuli](https://kiranparajuli.com.np) ~~strikes body~~ here some",
      "now with __underlined text__ within some ___underl_ined_text___",
    ])("should be deep tokenized", (line) => {
      const lines = [line]
      const lexerData = toHtml(lines)
      expect(lexerData).toMatchSnapshot()
    })
  })
  describe("bold", () => {
    it("should match the consecutive occurrences", () => {
      const lines = [
        "**first bold** **second bold** **third bold**"
      ]
      const lexerData = toHtml(lines)
      expect(lexerData).toMatchSnapshot()
    })
    it("should be deep tokenized", () => {
      const lines = [
        "**bold with `code` and *italics* and ~~strikes~~ and [link](https://kiranparajuli.com.np)**"
      ]
      const lexerData = toHtml(lines)
      expect(lexerData).toMatchSnapshot()
    })
  })
  describe("italics", () => {
    it("should match the consecutive occurrences", () => {
      const lines = [
        "*first italics* *second italics* *third italics*"
      ]
      const lexerData = toHtml(lines)
      expect(lexerData).toMatchSnapshot()
    })
    it("should be deep tokenized", () => {
      const lines = [
        "*italics with `code` and ~~strikes~~ and [link](https://kiranparajuli.com.np)*"
      ]
      const lexerData = toHtml(lines)
      expect(lexerData).toMatchSnapshot()
    })
  })
  describe("strikes", () => {
    it("should match the consecutive occurrences", () => {
      const lines = [
        "~~first strikes~~ ~~second strikes~~ ~~third strikes~~"
      ]
      const lexerData = toHtml(lines)
      expect(lexerData).toMatchSnapshot()
    })
    it("should be deep tokenized", () => {
      const lines = [
        "~~strikes with ~ gem and `code` and **bold** and *italics* and [link](https://kiranparajuli.com.np)~~"
      ]
      const lexerData = toHtml(lines)
      expect(lexerData).toMatchSnapshot()
    })
  })
  describe("code", () => {
    it("should match the consecutive occurrences", () => {
      const lines = [
        "`first code` `second code` `third code`"
      ]
      const lexerData = toHtml(lines)
      expect(lexerData).toMatchSnapshot()
    })
    it("should not be deep tokenized", () => {
      const lines = [
        "`code with ~ gem and **bold** and *italics* and [link](https://kiranparajuli.com.np)`"
      ]
      const lexerData = toHtml(lines)
      expect(lexerData).toMatchSnapshot()
    })
  })
  describe("link", () => {
    it("should match the consecutive occurrences", () => {
      const lines = [
        "[link](https://kiranparajuli.com.np) [link](https://kiranparajuli.com.np)"
      ]
      const lexerData = toHtml(lines)
      expect(lexerData).toMatchSnapshot()
    })
    it("should deep tokenized the link title", () => {
      const lines = [
        "[*italic* with **bold** and `code`](https://kiranparajuli.com.np)"
      ]
      const lexerData = toHtml(lines)
      expect(lexerData).toMatchSnapshot()
    })
  })
})
