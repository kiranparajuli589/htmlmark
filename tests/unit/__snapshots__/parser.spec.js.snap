// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser bold should be deep tokenized 1`] = `"<p><strong>bold with <code>code</code> and <em>italics</em> and <s>strikes</s> and <a href='https://kiranparajuli.com.np'>link</a></strong></p>"`;

exports[`Parser bold should match the consecutive occurrences 1`] = `"<p><strong>first bold</strong> <strong>second bold</strong> <strong>third bold</strong></p>"`;

exports[`Parser code should match the consecutive occurrences 1`] = `"<p><code>first code</code> <code>second code</code> <code>third code</code></p>"`;

exports[`Parser code should not be deep tokenized 1`] = `"<p><code>code with ~ gem and &lt;strong&gt;bold&lt;/strong&gt; and &lt;em&gt;italics&lt;/em&gt; and <a href='https://kiranparajuli.com.np'>link</a></code></p>"`;

exports[`Parser code should not panic with angles inside 1`] = `"<p>Pre-formatted code blocks are used for writing about programming or markup source code. Rather than forming normal paragraphs, the lines of a code block are interpreted literally. Markdown wraps a code block in both <code>&lt;pre&gt;</code> and <code>&lt;code&gt;</code> tags.</p>"`;

exports[`Parser codeblock should cope with multiple items 1`] = `
"
<pre><code class='language-js'>const a = 1</code></pre>
<p>some people are funny</p>"
`;

exports[`Parser codeblock should parse indent codeblock 1`] = `
"<p>This is a normal paragraph:</p>
<br>
<pre><code>This is a code block.</code></pre>

<br><p>Here is an example of AppleScript:</p>
<br>
<pre><code>tell application &quot;Foo&quot;
beep
end tell</code></pre>

<br><p>A code block continues until it reaches a line that is not indented (or the end of the article).</p>"
`;

exports[`Parser codeblock should parse multiple consequetive codeblocks 1`] = `
"
<pre><code class='language-js'>const a = 1</code></pre>

<pre><code class='language-js'>const b = 2</code></pre>
"
`;

exports[`Parser codeblock should parse the codeblock 1`] = `
"
<pre><code class='language-js'>const a = 1</code></pre>
"
`;

exports[`Parser codeblock should parse the codeblock with multiple lines 1`] = `
"
<pre><code class='language-js'>const a = 1
const b = 2</code></pre>
"
`;

exports[`Parser codeblock should parse the codeblock with multiple lines and a newline 1`] = `
"
<pre><code class='language-js'>const a = 1
const b = 2

const c = 3</code></pre>
"
`;

exports[`Parser codeblock should parse the codeblock without the language 1`] = `
"
<pre><code>const a = 1</code></pre>
"
`;

exports[`Parser comment should add a comment 1`] = `
"<p>one</p>
<!-- comment--><p>three</p>"
`;

exports[`Parser common tokens should parse the common tokens 1`] = `
"
<h1>Heading One Text</h1>"
`;

exports[`Parser common tokens should parse the common tokens 2`] = `
"
<h2>Heading Two Text</h2>"
`;

exports[`Parser common tokens should parse the common tokens 3`] = `
"
<h3>Heading Three Text</h3>"
`;

exports[`Parser common tokens should parse the common tokens 4`] = `
"
<h4>Heading Four Text</h4>"
`;

exports[`Parser common tokens should parse the common tokens 5`] = `
"
<h5>Heading Five Text</h5>"
`;

exports[`Parser common tokens should parse the common tokens 6`] = `
"
<h6>Heading Six Text</h6>"
`;

exports[`Parser common tokens should parse the common tokens 7`] = `
"
<!-- comment item text-->"
`;

exports[`Parser common tokens should parse the common tokens 8`] = `
"
<img src='image.png' alt='alt text'>"
`;

exports[`Parser common tokens should parse the common tokens with some space before 1`] = `
"
<h1>Heading One Text</h1>"
`;

exports[`Parser common tokens should parse the common tokens with some space before 2`] = `
"
<h2>Heading Two Text</h2>"
`;

exports[`Parser common tokens should parse the common tokens with some space before 3`] = `
"
<h3>Heading Three Text</h3>"
`;

exports[`Parser common tokens should parse the common tokens with some space before 4`] = `
"
<h4>Heading Four Text</h4>"
`;

exports[`Parser common tokens should parse the common tokens with some space before 5`] = `
"
<h5>Heading Five Text</h5>"
`;

exports[`Parser common tokens should parse the common tokens with some space before 6`] = `
"
<h6>Heading Six Text</h6>"
`;

exports[`Parser common tokens should parse the common tokens with some space before 7`] = `
"
<!-- comment item text-->"
`;

exports[`Parser common tokens should parse the common tokens with some space before 8`] = `
"
<img src='image.png' alt='alt text'>"
`;

exports[`Parser hr line should not allow multiple consecutive hr lines 1`] = `
"
<hr>"
`;

exports[`Parser hr line should parse the hr line 1`] = `
"
<hr>"
`;

exports[`Parser hr line should parse the hr line 2`] = `"<p>--</p>"`;

exports[`Parser hr line should parse the hr line 3`] = `
"
<hr>"
`;

exports[`Parser italics should be deep tokenized 1`] = `"<p><em>italics with <code>code</code> and <s>strikes</s> and <a href='https://kiranparajuli.com.np'>link</a></em></p>"`;

exports[`Parser italics should match the consecutive occurrences 1`] = `"<p><em>first italics</em> <em>second italics</em> <em>third italics</em></p>"`;

exports[`Parser link should deep tokenized the link title 1`] = `"<p><a href='https://kiranparajuli.com.np'><em>italic</em> with <strong>bold</strong> and <code>code</code></a></p>"`;

exports[`Parser link should match the consecutive occurrences 1`] = `"<p><a href='https://kiranparajuli.com.np'>link</a> <a href='https://kiranparajuli.com.np'>link</a></p>"`;

exports[`Parser list indent should break the list 1`] = `
"
<ul>
<li><p>item <strong>1</strong></p></li>
<li><p>item <a href='link-url'>link</a></p>
<ul>
<li><p>item 3 <code>code item</code></p>
<ul>
<li><p>item 4</p></li>
</ul></li>
<li><p>item 3 <code>code item</code></p></li>
</ul></li>
</ul>"
`;

exports[`Parser list should detect list indent 1`] = `
"
<ul>
<li><p>item <strong>1</strong></p></li>
<li><p>item <a href='link-url'>link</a></p></li>
<li><p>item 3 <code>code item</code></p></li>
</ul>"
`;

exports[`Parser list should parse the indented codeblock inside 1`] = `
"<p>To put a code block within a list item, the code block needs to be indented <em>twice</em> -- 8 spaces or two tabs:</p>
<ul>
<li><p>A list item with a code block:</p>
<br>
<pre><code>&lt;code goes here&gt;
&lt;code goes here&gt;
&lt;code goes here&gt;</code></pre>
</li>
</ul>
<br>
<h3>Code Blocks</h3>"
`;

exports[`Parser list should tokenize a valid ordered list 1`] = `
"
<ol>
<li><p>item <strong>1</strong></p></li>
<li><p>item <a href='link-url'>link</a></p></li>
<li><p>item 3 <code>code item</code></p></li>
</ol>"
`;

exports[`Parser list should tokenize a valid un-ordered list 1`] = `
"
<ul>
<li><p>item <strong>1</strong></p></li>
<li><p>item <a href='link-url'>link</a></p></li>
<li><p>item 3 <code>code item</code></p></li>
</ul>"
`;

exports[`Parser list should tokenize list combination 1`] = `
"
<ul>
<li><p>one</p></li>
<li><p>two</p></li>
</ul>
<ol>
<li><p>one</p></li>
<li><p>two</p></li>
</ol>
<ul>
<li><input type='checkbox'><p>c empty</p></li>
<li><input type='checkbox'><p>c checked</p></li>
</ul>
<ol>
<li><input type='checkbox'><p>c empty</p></li>
<li><input type='checkbox'><p>c checked</p></li>
</ol>"
`;

exports[`Parser newline should combine multiple consequest new lines to a single one 1`] = `
"<p>some plain text</p>
<br><p>some more plain text</p>"
`;

exports[`Parser newline should include the empty line after some content 1`] = `
"<p>some plain text</p>
<br>"
`;

exports[`Parser newline should include the empty line after some content 2`] = `
"<p>some plain text</p>
<br>"
`;

exports[`Parser newline should not include the top empty lines 1`] = `""`;

exports[`Parser newline should not include the top empty lines 2`] = `""`;

exports[`Parser newline should not include the top empty lines 3`] = `""`;

exports[`Parser paragraph should be deep tokenized 1`] = `"<p>some normal and <strong>bold with <em> gem</strong> but pure </em>italics<em> is alos there baby now <s>coming</s>hola amigons<s>strike</s> wooo lala what about blazing <strong>****</em></strong>here baby <s>~~~~~</s>baby <code>baby\`\`</code><a href='href'>[[[[[[[[[[[[[[l]]]]int</a><a href='href'>link</a><code>\`\`\`\`\`\`\`\`</code></p>"`;

exports[`Parser paragraph should be deep tokenized 2`] = `"<p>a paragraph of <u>words</u> <code>first code</code> normal text here <code>code body</code> <em>first italics</em> here me crying <em>italic body</em> here me crying <strong>first bolds</strong> some normal again <strong>bold body</strong> <a href='https://kiranparajuli.com.np'>Kiran Parajuli</a> <s>strikes body</s> here some</p>"`;

exports[`Parser paragraph should be deep tokenized 3`] = `"<p>now with <u>underlined text</u> within some <u>_underl_ined_text_</u></p>"`;

exports[`Parser quote should deep tokenize quote with multiple depth 1`] = `
"
<blockquote>
<blockquote><p>quote <strong>one</strong> with <em>two</em></p>
</blockquote></blockquote>"
`;

exports[`Parser quote should deep tokenize quote with multiple depth 2`] = `
"
<blockquote>
<blockquote>
<blockquote><p>quote <code>four</code> with <s>five</s></p>
</blockquote></blockquote></blockquote>"
`;

exports[`Parser quote should deep tokenize quote with multiple depth 3`] = `
"
<blockquote>
<blockquote>
<blockquote>
<blockquote><p>quote <a href='link-url'>link-title</a> with <em>two</em></p>
</blockquote></blockquote></blockquote></blockquote>"
`;

exports[`Parser quote should deep tokenize quote with zero depth 1`] = `
"
<blockquote><p>quote <strong>one</strong> with <em>two</em></p>
</blockquote>"
`;

exports[`Parser quote should deep tokenize quote with zero depth 2`] = `
"
<blockquote><p>quote <code>four</code> with <s>five</s></p>
</blockquote>"
`;

exports[`Parser quote should deep tokenize quote with zero depth 3`] = `
"
<blockquote><p>quote <a href='link-url'>link-title</a> with <em>two</em></p>
</blockquote>"
`;

exports[`Parser quote should detect the circular quote 1`] = `
"
<blockquote><p>one</p>
<blockquote><p>two</p>
<blockquote><p>three</p>
<blockquote><p>four</p>
<blockquote><p>five</p>
</blockquote>
<br><p>four</p>
</blockquote>
<br><p>three</p>
</blockquote>
<br><p>two</p>
</blockquote>
<br><p>one</p>
</blockquote>"
`;

exports[`Parser quote should find the list and codeblock inside 1`] = `
"
<blockquote>
<h2>This is a header.</h2>
<br>
<ol>
<li><p>This is the first list item.</p></li>
<li><p>This is the second list item.</p></li>
</ol>
<br><p>Here's some example code:</p>
<br>
<pre><code>return shell_exec(&quot;echo $input | $markdown_script&quot;);</code></pre>

</blockquote>"
`;

exports[`Parser quote should parse multiline quote with the same depth and indent 1`] = `
"
<blockquote>
<blockquote>
<blockquote><p>zero f one f</p>
<br>
</blockquote>
</blockquote><p>one</p>
<h1>two</h1>
<blockquote><p>three</p>
<blockquote><p>four</p>
<blockquote><p>d-five</p>
</blockquote>
</blockquote>
<h1>five</h1>
<h2>six</h2>
<blockquote>
<blockquote><p>seven simple para</p>
</blockquote></blockquote>
</blockquote>
</blockquote>"
`;

exports[`Parser quote should parse the quote with list inside 1`] = `
"
<blockquote>
<ul>
<li><p>one</p></li>
<li><p>two</p></li>
</ul>
</blockquote>"
`;

exports[`Parser strikes should be deep tokenized 1`] = `"<p><s>strikes with ~ gem and <code>code</code> and <strong>bold</strong> and <em>italics</em> and <a href='https://kiranparajuli.com.np'>link</a></s></p>"`;

exports[`Parser strikes should match the consecutive occurrences 1`] = `"<p><s>first strikes</s> <s>second strikes</s> <s>third strikes</s></p>"`;

exports[`Parser table is not a table false separator 1`] = `"<p>| column 1 | column 2 | | - | - | | row 1 c1 | row 1 c2 |</p>"`;

exports[`Parser table is not a table not consistent cell count 1`] = `"<p>| column 1 | column 2 | | --- | | row 1 c1 | row 1 c2 |</p>"`;

exports[`Parser table is not a table only header 1`] = `"<p>| column 1 | column 2 |</p>"`;

exports[`Parser table is not a table only header and separator 1`] = `"<p>| column 1 | column 2 | |---|---|</p>"`;

exports[`Parser table is not a table other tokens in between 1 1`] = `
"<p>| column 1 | column 2 |</p>
<br><p>|---|---| | row 1 c1 | row 1 c2 |</p>"
`;

exports[`Parser table is not a table other tokens in between 2 1`] = `
"<p>| column 1 | column 2 | |---|---|</p>
<br><p>| row 1 c1 | row 1 c2 |</p>"
`;

exports[`Parser table is not a table other tokens in between 3 1`] = `
"
<table>
  <thead>
    <tr>
    	<th>column 1</th><th>column 2</th>
		</tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr>
  </tbody>
</table>
<br><p>| row 2 c1 | row 2 c2 |</p>"
`;

exports[`Parser table table cell count cell count should break the table 1 1`] = `
"
<table>
  <thead>
    <tr>
    	<th>column 1</th><th>column 2</th>
		</tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr>
  </tbody>
</table><p>| row 2 c1 |</p>"
`;

exports[`Parser table table cell count cell count should break the table 2 1`] = `
"
<table>
  <thead>
    <tr>
    	<th>column 1</th><th>column 2</th>
		</tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
    	<th>column 1</th>
		</tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td></tr>
  </tbody>
</table>"
`;

exports[`Parser table table heading-body separator should parse a table 1`] = `
"
<table>
  <thead>
    <tr>
    	<th>column 1</th><th>column 2</th>
		</tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr><tr><td>row 2 c1</td><td>row 2 c2</td></tr>
  </tbody>
</table>"
`;

exports[`Parser table table heading-body separator should parse a table 2`] = `
"
<table>
  <thead>
    <tr>
    	<th>column 1</th><th>column 2</th>
		</tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr><tr><td>row 2 c1</td><td>row 2 c2</td></tr>
  </tbody>
</table>"
`;

exports[`Parser table table indent acceptable indent should not break the table 1 1`] = `
"
<table>
  <thead>
    <tr>
    	<th>column 1</th><th>column 2</th>
		</tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr><tr><td>row 2 c1</td><td>row 2 c2</td></tr>
  </tbody>
</table>"
`;

exports[`Parser table table indent different indent header -> separator 1`] = `"<p>| column 1 | column 2 | |---|---| | row 1 c1 | row 1 c2 |</p>"`;

exports[`Parser table table indent different indent header, separator -> body row 1`] = `
"
<pre><code>| column 1 | column 2 |
|---|---|</code></pre>
<p>| row 1 c1 | row 1 c2 |</p>"
`;

exports[`Parser table table indent indent should break the table 2 1`] = `
"
<table>
  <thead>
    <tr>
    	<th>column 1</th><th>column 2</th>
		</tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
    	<th>column 1</th><th>column 2</th>
		</tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr>
  </tbody>
</table>"
`;

exports[`Parser table table indent the indented table 1`] = `
"
<table>
  <thead>
    <tr>
    	<th>column 1</th><th>column 2</th>
		</tr>
  </thead>
  <tbody>
    <tr><td>row 1 c1</td><td>row 1 c2</td></tr>
  </tbody>
</table>"
`;
